pkg/controller/networkpolicy/controller_query_perf_test.go

```diff
@@ -0,0 +1,241 @@
+		wg.Done()
+	}()
+	// create controller
+	objs := toRunTimeObjects(namespaces, networkPolicies, pods)
+	_, querier := makeControllerAndEndpointQueryReplier(objs...)
+	// Everything is ready, now start timing.
+	start := time.Now()
+	// track execution time by calling query endpoint 1000 times on random pods
+	for i := 0; i < 1000; i++ {
+		pod, namespace := pods[i].Name, pods[i].Namespace
+		response, err := querier.QueryNetworkPolicies(namespace, pod)
+		require.Equal(t, err, nil)
+		require.Equal(t, len(response.Endpoints[0].Policies), responseLength)
+	}
+	// Stop tracking go routines
+	stopCh <- struct{}{}
+	// Minus the idle time to get the actual execution time.
+	executionTime := time.Since(start)
+	if executionTime > maxExecutionTime {
+		t.Errorf("The actual execution time %v is greater than the maximum value %v", executionTime, maxExecutionTime)
+	}
+
+	// Block until all statistics are done.
+	wg.Wait()
+
+	t.Logf(`Summary metrics:
+NAMESPACES   PODS    NETWORK-POLICIES    TIME(s)    MEMORY(M)    
+%-12d %-7d %-19d %-10.2f %-12d 
+`, len(namespaces), len(pods), len(networkPolicies), float64(executionTime)/float64(time.Second), maxAlloc/1024/1024)
+}
+
+func genUniqueRuntimeObjects(numObjects int) (namespaces []*v1.Namespace, networkPolicies []*networkingv1.NetworkPolicy, pods []*v1.Pod) {
+	namespaceID := rand.String(8)
+	namespaces, networkPolicies, pods = make([]*v1.Namespace, 0), make([]*networkingv1.NetworkPolicy, 0), make([]*v1.Pod, 0)
+	for i := 0; i < numObjects; i++ {
+		namespace := &v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{Name: namespaceID, Labels: map[string]string{"app": namespaceID}},
+		}
+		networkPolicy := &networkingv1.NetworkPolicy{
+			ObjectMeta: metav1.ObjectMeta{Namespace: namespaceID, Name: "np-" + string(i), UID: types.UID(uuid.New().String())},
+			Spec: networkingv1.NetworkPolicySpec{
+				PodSelector: metav1.LabelSelector{MatchLabels: map[string]string{"app-1": "scale-1"}},
+				PolicyTypes: []networkingv1.PolicyType{networkingv1.PolicyTypeIngress, networkingv1.PolicyTypeEgress},
+				Ingress: []networkingv1.NetworkPolicyIngressRule{
+					{
+						From: []networkingv1.NetworkPolicyPeer{
+							{
+								PodSelector: &metav1.LabelSelector{
+									MatchLabels: map[string]string{"app-1": "scale-1"},
+								},
+							},
+						},
+					},
+				},
+			},
+		}
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: namespaceID,
+				Name: "pod" + string(i),
+				UID: types.UID(uuid.New().String()),
+				Labels: map[string]string{"app-1": "scale-1"},
+			},
+			Spec:       v1.PodSpec{NodeName: getRandomNodeName()},
+			Status:     v1.PodStatus{PodIP: getRandomIP()},
+		}
+		namespaces = append(namespaces, namespace)
+		networkPolicies = append(networkPolicies, networkPolicy)
+		pods = append(pods, pod)
+	}
+	return namespaces, networkPolicies, pods
+}
+
+/*
+Note, this is performance data relevant to a previous implementation of QueryNetworkPolicy
```

Comment:
Oh I see that you have the numbers here. Are these up-to-date: does the "with indexing" case include both the AddressGroup and AppliedToGroup indexer? Do they account for the extra change we had to make the other day (add Pod references for AddressGroups)?

Reply:
Good question, the "with indexing" section is up to date. However, I would need to make a new implementation for "without indexing" in order to properly benchmark it. I am not sure if that is necessary though. I added "++" following the numbers to indicate that the numbers will likely be higher following any implementation without indexing.

Reply:
Why would the "without indexing" case change? That was before we added any indexers

Reply:
It doesn't take into account egress-ingress rules.

Reply:
If it were to support egress-ingress rules the time + memory would be different.

Reply:
I'm really mostly interested in the memory footprint of adding the indexers. I think that's orthogonal from your query logic.

Reply:
We will just need to have these numbers handy

Reply:
If we want to have a memory footprint for querying without indexing, I will need to find a way to implement the rules functionality without using indexing, which may take some time. However, we can see relative memory footprint for initialization with/ without indexing, which I could add in without re-implementing rule functionality.

Reply:
Initialization is what "accounts" for the memory footprint: the stores & the indexers consume memory. The query operation itself should not matter.